src/utils/AuthContext.tsx

```diff
@@ -1,46 +1,87 @@
 import {
   AuthError,
-  AuthResponse,
   Session,
   User,
   UserAttributes,
-  UserResponse,
 } from '@supabase/supabase-js';
 import React, {
   createContext,
   useContext,
   useEffect,
   useMemo,
-  useState,
+  useReducer,
 } from 'react';
 
 import supabase from './supabase';
 
+type AuthContextAction =
+  | { type: 'LOADING' }
+  | { type: 'CLEAR_ERROR' }
+  | { type: 'SIGN_UP' }
+  | { type: 'SIGN_IN_WITH_EMAIL' }
+  | { type: 'VERIFY_OTP' }
+  | { type: 'RESEND_VERIFICATION' }
+  | { type: 'RESET_PASSWORD' }
+  | { type: 'REFRESH_SESSION'; session: Session | null }
+  | { type: 'UPDATE_USER'; user: User | null }
+  | { type: 'SIGN_OUT' };
+
+export type AuthDispatch = React.Dispatch<AuthContextAction>;
+
 export interface AuthState {
   session: Session | null;
   user: User | null;
   isLoading: boolean;
-  signIn: (newSession: Session | null) => void;
-  signUp: (email: string, password: string) => Promise<AuthResponse>;
-  signInWithEmail: (email: string, password: string) => Promise<AuthResponse>;
-  verifyOtp: (email: string, token: string) => Promise<AuthResponse>;
-  resendVerification: (email: string) => Promise<AuthResponse>;
-  resetPassword: (email: string) => Promise<
-    | {
-        data: object;
-        error: null;
-      }
-    | {
-        data: null;
-        error: AuthError;
-      }
-  >;
-  updateUser: (attributes: UserAttributes) => Promise<UserResponse>;
-  signOut: () => Promise<void>;
+  dispatch: AuthDispatch;
 }
 
 const AuthContext = createContext({} as AuthState);
 
+export const useAuthReducer = () =>
+  useReducer(
+    (prevState: AuthState, action: AuthContextAction) => {
+      switch (action.type) {
+        case 'SIGN_UP':
+        case 'SIGN_IN_WITH_EMAIL':
+        case 'VERIFY_OTP':
+        case 'RESEND_VERIFICATION':
+        case 'RESET_PASSWORD':
+        case 'UPDATE_USER':
+          return {
+            ...prevState,
+            isLoading: false,
+          };
+        case 'SIGN_OUT':
+          return {
+            ...prevState,
+            user: null,
+            session: null,
+            isLoading: false,
+          };
+        case 'REFRESH_SESSION':
+          return {
+            ...prevState,
+            session: action.session,
+            user: action.session ? action.session?.user : null,
+            isLoading: false,
+          };
+        case 'LOADING':
+          console.log('loading');
+          return { ...prevState, isLoading: true };
+        case 'CLEAR_ERROR':
```

Comment:
i don't rly see this being used anywhere, so i might recommend to remove it. if yall decide to handle errors/clearing errors with context then you can maybe add it back in a later PR.